design/batch_retrieval_design.md

```diff
@@ -2,19 +2,140 @@
-| 3    | Stay as is                                                   | No additional work needed                                    | Bad UX                                                       | -                     |
+|      | Solution                                                     | Pros                                                         | Cons                                               | Effort (T-shirt size) |
+| ---- | ------------------------------------------------------------ | ------------------------------------------------------------ | -------------------------------------------------- | --------------------- |
+| 1    | *Client side:* <br />For each K8s Secret, perform a new Batch request | - No server side changes / no breaking changes               | Load on server with extra calls (*)                | S, 5 days             |
+| 2    | *Server side:*<br />Update Batch retrieval to return list of variables **and** their response (success/failure) (**) | - Will help us during rotation for Milestone 2<br />- Best practice / straightforward design for batch endpoint | - Requires both client/server implementation<br /> | M, 10 days            |
+| 3    | Stay as is                                                   | No additional work needed                                    | Bad UX                                             | -                     |
 
 (*) Fallback solution: use this solution as a safety net, only if the original Batch Retrieval request fails
 
 (**) This solution can be broken up in two: 1. Create a new API endpoint or 2. use existing one (return 409 response)
 
+*Decision:* We decided to go with #2 and create our own API. The new batch retrieval API endpoint will not be forward facing and be for **internal use only**. 
+
+### Design
+
+#### Secrets Provider Job flow
+
+As a small recap, the following explains the Secrets Provider Job flow
+
+1. The `cyberark-secrets-provider-for-k8s` runs as a Job and authenticates with the Conjur/DAP Follower using the Kubernetes Authenticator (`authn-k8s`).
+
+2. The `cyberark-secrets-provider-for-k8s` reads all Kubernetes secrets required by the applications in the same namespace.
+
+3. For each mapped Kubernetes secret, 
+
+   a. The `cyberark-secrets-provider-for-k8s` performs a Batch Retrieval request and retrieves Conjur/DAP secrets.
+
+   b. It updates the Kubernetes secret with the Conjur/DAP secret value.
+
+4. The `cyberark-secrets-provider-for-k8s` Job is completed.
+
+The new Batch Retrieval API will effect #2a. 
+
+### URI
+
+##### POST /secrets/batch
+
+This method retuns a list of secrets, their status code, and their value if an error did not occur. 
+
+```
+POST /secrets/batch
```

Comment:
is this the new API endpoint? so we'll now have:
- endpoint `/secrets` that performs a batch retrieval and fails if one secrets fails
- endpoint `/secrets/batch` that performs a batch retrieval and does not fail if one secrets fails

did i get it right?

Reply:
will both APIs have the query param `variable_ids`?

Reply:
Exactamundo!

Reply:
i think that this is quite drunk. i would think that the difference should be in the url itself. both `/secrets` and `/secrets/batch` perform batch retrieval for secrets. 

i will add another comment with my thoughts on this.

Reply:
I'd +1 to this. Maybe something like `/secrets/batch-v2` or `/secrets/batch/v2` would make more sense?

Reply:
We might want to deprecate the old batch api. Does it still make sense to add the `v2`?

Reply:
do we have a v2 concept there for other REST APIs in conjur ?
If not - we should not add v2 only for one specific API. V2 is usually used when rewriting all APIs to some new concept

Reply:
changed to `POST /secrets` as discussed in the meeting